Gems/ScriptCanvas/Code/Include/ScriptCanvas/Serialization/GraphDataSerializer.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Serialization/Json/JsonSerialization.h>
+#include <ScriptCanvas/Core/SerializationListener.h>
+#include <ScriptCanvas/Core/GraphData.h>
+#include <ScriptCanvas/Serialization/GraphDataSerializer.h>
+
+using namespace ScriptCanvas;
+
+namespace GraphDataSerializerCpp
+{
+    void CollectNodes(const GraphData::NodeContainer& container, SerializationListeners& listeners)
+    {
+        for (auto& nodeEntity : container)
+        {
+            if (nodeEntity)
+            {
+                if (auto listener = azrtti_cast<SerializationListener*>(AZ::EntityUtils::FindFirstDerivedComponent<Node>(nodeEntity)))
+                {
+                    listeners.push_back(listener);
+                }
+            }
+        }
+    }
+}
+
+namespace AZ
+{
+    AZ_CLASS_ALLOCATOR_IMPL(GraphDataSerializer, SystemAllocator, 0);
+
+    JsonSerializationResult::Result GraphDataSerializer::Load
+        ( void* outputValue
+        , [[maybe_unused]] const Uuid& outputValueTypeId
+        , const rapidjson::Value& inputValue
+        , JsonDeserializerContext& context)
+    {
+        namespace JSR = JsonSerializationResult;
+
+        AZ_Assert(outputValueTypeId == azrtti_typeid<GraphData>()
+            , "RuntimeVariableSerializer Load against output typeID that was not GraphData");
+        AZ_Assert(outputValue, "RuntimeVariableSerializer Load against null output");
+        context.GetMetadata().Add(SerializationListeners());
+        JSR::ResultCode result(JSR::Tasks::ReadField);
+        result.Combine(ContinueLoading(outputValue, outputValueTypeId, inputValue, context, ContinuationFlags::NoTypeSerializer));
+        auto listeners = context.GetMetadata().Find<SerializationListeners>();
+        AZ_Assert(listeners, "Failed to create SerializationListeners");
+        GraphDataSerializerCpp::CollectNodes(reinterpret_cast<GraphData*>(outputValue)->m_nodes, *listeners);
+
+        for (auto listener : *listeners)
+        {
+            listener->OnDeserialize();
+        }
+
+        reinterpret_cast<GraphData*>(outputValue)->OnDeserialized();
```

Comment:
Consider doing this post-processing after deserialization has completed, so after the Json Serialization is done. This keeps a cleaner separation between the Json Serialization part and other processing. (This is in particular relevant to Prefabs as one of the things being considered is multi-threaded processing.) The `SerializationListeners` can also be added to the meta-data through the settings passed into the Json Serialization. (Same applies to the `Store` function.)

Reply:
Given that this will only occur at the top level of the asset, that will simply mean moving the call to the asset handler, which is fine. I do want to know, at this point, I consider the JSON serialization to be "done". Or is there more than will happen after this point?

Reply:
When the top level API level such as `AZ::JsonSerialization::Load(...);` returns it's considered "done" with that task. Any custom json serializers added are considered part of the processing so when they complete a call then the task is "not done". The idea behind this is that the Json Serialization captures all information it needs upfront and from there on out no external mutable data is accessed. This provides guarantees that the Json Serialization can be run in parallel. This is why it's better to use the meta-data to store any information for later processing instead of immediately executing it in the serializer. (However there's no way to guarantee any of custom json serializers don't misbehave and still mutates data outside the Json Serialization so your mileage may vary.)